symengine/serialize-cereal.h

```diff
@@ -25,6 +25,113 @@
 namespace SymEngine
 {
 
+template <class Archive>
+class RCPBasicAwareOutputArchive : public Archive
+{
+    using Archive::Archive;
+
+public:
+    void save_rcp_basic(const RCP<const Basic> &ptr)
+    {
+        uintptr_t addr = (uintptr_t)(void *)ptr.get();
+        (*this)(addr);
+
+        auto id = _addresses.find(addr);
+        int8_t first_seen = (id == _addresses.end());
+        (*this)(first_seen);
+
+        if (not first_seen) {
+            return;
+        }
+        TypeID type_code = ptr->get_type_code();
+        save_typeid(*this, type_code);
+        switch (type_code) {
+#define SYMENGINE_ENUM(type, Class)                                            \
+    case type:                                                                 \
+        save_basic(*this, static_cast<const Class &>(*ptr));                   \
+        break;
+#include "symengine/type_codes.inc"
+#undef SYMENGINE_ENUM
+            default:
+                save_basic(*this, *ptr);
+        }
+        _addresses.insert(addr);
+    }
+
+private:
+    std::set<uintptr_t> _addresses;
+    //! Overload the rtti function to enable dynamic_cast
+    void rtti(){};
+};
+
+template <class Archive>
+class RCPBasicAwareInputArchive : public Archive
+{
+    using Archive::Archive;
+
+public:
+    template <class T>
+    RCP<const T> load_rcp_basic()
+    {
+        try {
+            uintptr_t addr;
+            (*this)(addr);
+
+            int8_t first_seen;
+            (*this)(first_seen);
```

Comment:
can we add an assert below this line so that we in debug build make sure that we are not reading garbage?
something along the lines of:
`assert(first_seen == 0 || first_seen == 1);`

Reply:
That would fail, as the test that failed was checking that a corrupted input doesn't segfault.

Reply:
Ah, now I see. Then maybe the handling of `bool` was fine all along?
Then we don't want a conditional throw of a `SerializationError` instead of assert either, since that would make our test coverage worse? I'm fine with this as is then.

Reply:
We could do a SerializationError to indicate a corrupt input.

Reply:
Alright, I'll leave the descision to you, I'm fine with both approaches :)